File: app/src/main/java/com/neworld/youyou/activity/SettingActivity.java
package com.neworld.youyou.activity;

import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.CardView;
import android.util.Base64;
import android.view.View;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.LinearLayout;

import com.neworld.youyou.R;
import com.neworld.youyou.bean.ReturnStatus;
import com.neworld.youyou.manager.MyApplication;
import com.neworld.youyou.manager.NetManager;
import com.neworld.youyou.utils.GsonUtil;
import com.neworld.youyou.utils.Sputil;

public class SettingActivity extends AppCompatActivity implements View.OnClickListener {

    private ImageView ivClose;
    private Button btQuit;
    private LinearLayout linearLayout;
    private String userId;
    private MyApplication application;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_me_settings);
        if (application == null) {
            application = (MyApplication) getApplication();
        }
        initUser();
        initView();
    }

    private void initUser() {
        userId = Sputil.getString(SettingActivity.this, "userId", "");
    }

    private void initView() {
        ivClose = $(R.id.iv_close);
        btQuit = $(R.id.bt_quit);
        linearLayout = $(R.id.ll_setting);
        ivClose.setOnClickListener(this);
        btQuit.setOnClickListener(this);

        $(R.id.black_list).setOnClickListener(this);
        $(R.id.address_manager).setOnClickListener(this);
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.iv_close:
                finish();
                break;
            case R.id.black_list:
                startActivity(new Intent(SettingActivity.this, BlackNameActivity.class));
                break;
            case R.id.bt_quit:
                tanDialog();
                break;
            case R.id.address_manager:
                startActivity(new Intent(this, AddressActivity.class).putExtra("fromSetting", true));
                break;
        }
    }

    //弹出dialog
    private void tanDialog() {
        final AlertDialog.Builder ad = new AlertDialog.Builder(SettingActivity.this);
        ad.setMessage("确定退出登录吗");
        ad.setPositiveButton("确定",
                new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        quit();
                        dialog.dismiss();
                    }
                });
        ad.setNegativeButton("取消",
                new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        dialog.dismiss();
                    }
                });
        ad.show();
    }

    private void quit() {
        new Thread(new Runnable() {
            @Override
            public void run() {
                String base64 = Base64.encodeToString(("{\"userId\":\"" + userId + "\"}").getBytes(), Base64.DEFAULT);
                String replace = base64.replace("\n", "");
                String content = NetManager.getInstance().getContent(replace, "152");
                if (content != null && content.length() > 0) {
                    ReturnStatus returnStatus = GsonUtil.parseJsonToBean(content, ReturnStatus.class);
                    if (returnStatus != null && returnStatus.getStatus() == 0) {
                        Sputil.saveString(SettingActivity.this, "userId", "");
                        startActivity(new Intent(SettingActivity.this, LoginActivity.class));
                        application.removeALLActivity_();
                        SettingActivity.this.finish();
                    }
                }
            }
        }).start();
    }

    private <T extends View> T $(int res) {
        return (T) findViewById(res);
    }
}

File: app/src/main/java/com/neworld/youyou/activity/SettingActivity.kt
package com.neworld.youyou.activity

import android.content.Intent
import android.os.Bundle
import android.view.View

import com.neworld.youyou.R
import com.neworld.youyou.add.base.Activity
import com.neworld.youyou.bean.ReturnStatus
import com.neworld.youyou.manager.MyApplication
import com.neworld.youyou.showSnackBar
import com.neworld.youyou.utils.*
import com.neworld.youyou.view.mview.parents.Topics
import kotlinx.android.synthetic.main.activity_me_settings.*

class SettingActivity : Activity(), View.OnClickListener {
	
	private var userId by preference("userId", "")
	private var mApplication: MyApplication? = null
	private var homeCache  = ""
	private var cacheArray: Array<String>? = null

    override fun getContentLayoutId() = R.layout.activity_me_settings

    override fun initArgs(bundle: Bundle?): Boolean {
        if (mApplication == null) mApplication = application as MyApplication
		cacheArray = bundle?.getStringArray("typeArray")
		if (cacheArray != null && cacheArray!!.isNotEmpty()) {
			val edit = getPefStorage()
			cacheArray!!.forEach {
				val key = "${Topics.CACHE_KEY}$it"
				homeCache += edit.getString(key, "")
			}
		}
        homeCache = MyEventBus.INSTANCE.postEvent(Topics.CACHE_LENGTH, "")
                .fold("") { total, next -> total + (next ?: "") }
        return super.initArgs(bundle)
    }

	override fun initWidget() {
        iv_close.setOnClickListener(this)
        bt_quit.setOnClickListener(this)

//        black_list.setOnClickListener(this)
//        address_manager.setOnClickListener(this)

//        message_control.setOnClickListener(this)

		cache_size.text = convertSize()
		_clear.setOnClickListener {
			if (homeCache.isEmpty()) {
				showToast("已经清空啦！")
				return@setOnClickListener
			}
			try {
                val array = MyEventBus.INSTANCE.postEvent(Topics.CLEAR_CACHE)
                if ("1" !in array) {
                    homeCache = MyEventBus.INSTANCE.postEvent(Topics.CACHE_LENGTH, "")
                            .fold("") { total, next -> total + (next ?: "") }
                    cache_size.text = convertSize()
                    showToast("出现未知错误…")
                } else {
                    homeCache = ""
                    cache_size.text = convertSize()
                    showToast("清理成功！")
                }
			} catch (e: Exception) {
				showToast("清理失败！如多次出现请到用户反馈处反馈此问题 : $e")
			}
		}
    }

	override fun onClick(v: View) {
		when (v.id) {
			R.id.iv_close -> finish()
			R.id.black_list -> startActivity(Intent(this@SettingActivity, BlackNameActivity::class.java))
			R.id.bt_quit -> displayDialog(this, "确定退出登录吗", this@SettingActivity::quit)
			R.id.address_manager -> startActivity(Intent(this, AddressActivity::class.java).putExtra("fromSetting", true))
			R.id.message_control -> {
			// 消息界面: 195 userId (暂时没做)
                showToast("open the message center")
			}
		}
	}

	private fun quit() {
        response(::success, "152", "\"userId\":\"$userId\"")
	}
	
	private fun success(t: ReturnStatus) {
		if (t.status == 0) {
			userId = ""
			startActivity(Intent(this@SettingActivity, LoginActivity::class.java))
			mApplication?.removeALLActivity_()
			finish()
		} else {
			showSnackBar(_parent, "网络出错, 退出失败(如果持续出现此问题请到用户反馈处反馈)", 2000)
		}
	}

	/*override fun onStart() {
		super.onStart()
        *//*doAsync { // 消息提示
            val response = NetBuild.getResponse("\"userId\":\"$userId\"", 194) ?: "null"
            uiThread { _hint.visibility = if ("1" in response) View.VISIBLE else View.GONE }
        }*//*
	}*/

	private fun convertSize(): CharSequence {
		val p0 = (homeCache.toByteArray().size.toFloat() / 1024f / 1024f).toString()
		return if (p0.length > 4) "${p0.subSequence(0, 4)} Mb" else "$p0 Mb"
	}
}
